File: salary_calculator.py
import pandas as pd
import numpy as np
from datetime import datetime

class SalaryCalculator:
    def __init__(self, components):
        self.components = components

    def calculate_monthly_gross(self, annual_gross):
        """Calculate monthly gross from annual gross."""
        return round(annual_gross / 12, 2)

    def calculate_working_days_ratio(self, start_date, end_date):
        """Calculate the ratio of weekdays worked in the month."""
        start = pd.to_datetime(start_date)
        end = pd.to_datetime(end_date)

        # Get the month range
        month_start = pd.Timestamp(end.year, end.month, 1)
        month_end = month_start + pd.offsets.MonthEnd(1)

        # Create date ranges for worked days and total month days
        worked_days_range = pd.date_range(start=start, end=end)
        month_days_range = pd.date_range(start=month_start, end=month_end)

        # Count weekdays (Monday = 0, Sunday = 6)
        worked_weekdays = len([d for d in worked_days_range if d.weekday() < 5])
        total_weekdays = len([d for d in month_days_range if d.weekday() < 5])

        return round(worked_weekdays / total_weekdays, 2)

    def calculate_components(self, monthly_gross, working_days_ratio=1):
        """Calculate individual salary components."""
        return {
            component: round((monthly_gross * (percentage / 100)) * working_days_ratio, 2)
            for component, percentage in self.components.items()
        }

    def calculate_pension(self, basic, transport, housing, contract_type, monthly_gross, voluntary_pension=0):
        """Calculate pension contributions after proration."""
        if contract_type.strip().upper() == 'CONTRACT' or monthly_gross < 30000:
            return {
                'employee_pension': 0.00,
                'employer_pension': 0.00,
                'voluntary_pension': 0.00,
                'total_pension': 0.00
            }
        
        pensionable_base = basic + transport + housing
        employee_pension = round(0.08 * pensionable_base, 2)  # Fixed 8%
        employer_pension = round(0.10 * pensionable_base, 2)  # Fixed 10%
        voluntary = round(voluntary_pension, 2)
        total_pension = employee_pension + employer_pension + voluntary
        
        return {
            'employee_pension': employee_pension,
            'employer_pension': employer_pension,
            'voluntary_pension': voluntary,
            'total_pension': total_pension
        }

    def calculate_cra(self, gross_pay, pension):
        """Calculate Consolidated Relief Allowance (CRA)."""
        # Gross pay after pension deduction
        gross_after_pension = gross_pay - pension

        # Calculate 20% of gross after pension
        cra_percentage = round(0.2 * gross_after_pension, 2)

        # Calculate 1% of gross after pension and compare with monthly 200,000/12
        minimum_relief = round(max(0.01 * gross_after_pension, 200000 / 12), 2)

        return round(cra_percentage + minimum_relief, 2)

    def calculate_paye(self, taxable_pay):
        """Calculate PAYE tax using progressive tax bands."""
        tax_bands = [
            (300000, 0.07),
            (300000, 0.11),
            (500000, 0.15),
            (500000, 0.19),
            (1600000, 0.21),
            (float('inf'), 0.24)
        ]

        annual_taxable = taxable_pay * 12
        total_tax = 0
        remaining_income = annual_taxable

        for band, rate in tax_bands:
            if remaining_income <= 0:
                break
            taxable_in_band = min(band, remaining_income)
            total_tax += taxable_in_band * rate
            remaining_income -= band

        return round(total_tax / 12, 2)

    def process_employee(self, row):
        """Process salary calculations for a single employee."""
        # Calculate working days ratio
        working_ratio = self.calculate_working_days_ratio(row['START DATE'], row['END DATE'])

        # Calculate prorated monthly gross
        monthly_gross = self.calculate_monthly_gross(row['ANNUAL GROSS PAY'])
        prorated_monthly_gross = round(monthly_gross * working_ratio, 2)

        # Get reimbursements and other deductions
        reimbursements = float(row.get('Reimbursements', 0))
        other_deductions = float(row.get('Other Deductions', 0))

        # Calculate prorated components
        components = self.calculate_components(monthly_gross, working_ratio)

        # Calculate pension contributions
        pension_details = self.calculate_pension(
            components['BASIC'],
            components['TRANSPORT'],
            components['HOUSING'],
            row['Contract Type'],
            prorated_monthly_gross,
            float(row.get('VOLUNTARY_PENSION', 0))
        )

        # Calculate adjusted gross income for CRA
        statutory_deductions = pension_details['employee_pension'] + pension_details['voluntary_pension']
        adjusted_gross = prorated_monthly_gross - statutory_deductions
        
        # Calculate CRA using adjusted gross income
        cra = self.calculate_cra(adjusted_gross, 0)  # Pension already deducted from gross

        # Calculate taxable pay
        taxable_pay = round(adjusted_gross - cra, 2)

        # Calculate PAYE tax
        paye_tax = self.calculate_paye(taxable_pay)

        # Calculate total deductions and net pay
        total_deductions = round(
            paye_tax + 
            pension_details['employee_pension'] + 
            pension_details['voluntary_pension'] + 
            other_deductions, 2
        )
        net_pay = round(prorated_monthly_gross - total_deductions + reimbursements, 2)

        # Calculate total tax relief (sum of all tax-deductible amounts)
        total_tax_relief = round(cra + pension_details['employee_pension'] + pension_details['voluntary_pension'], 2)
        
        ordered_result = {
            'Account Number': row['Account Number'],
            'STAFF ID': row['STAFF ID'],
            'Email': row['Email'],
            'NAME': row['NAME'],
            'DEPARTMENT': row['DEPARTMENT'],
            'JOB TITLE': row['JOB TITLE'],
            'Contract Type': row['Contract Type'],
            'ANNUAL GROSS PAY': row['ANNUAL GROSS PAY'],
            'MONTHLY_GROSS': monthly_gross,
            'START DATE': row['START DATE'],
            'END DATE': row['END DATE'],
            'WORKING_DAYS_RATIO': working_ratio,
            'PRORATED_MONTHLY_GROSS': prorated_monthly_gross,
            'COMP_BASIC': components['BASIC'],
            'COMP_TRANSPORT': components['TRANSPORT'],
            'COMP_HOUSING': components['HOUSING'],
            'COMP_UTILITY': components['UTILITY'],
            'COMP_MEAL': components.get('MEAL', 0),
            'COMP_CLOTHING': components.get('CLOTHING', 0),
            'CRA': cra,
            'MANDATORY_PENSION': pension_details['employee_pension'],
            'VOLUNTARY_PENSION': pension_details['voluntary_pension'],
            'EMPLOYER_PENSION': pension_details['employer_pension'],
            'TAX_RELIEF': total_tax_relief,
            'TAXABLE_PAY': taxable_pay,
            'PAYE_TAX': paye_tax,
            'OTHER_DEDUCTIONS': other_deductions,
            'REIMBURSEMENTS': reimbursements,
            'TOTAL_DEDUCTIONS': total_deductions,
            'NET_PAY': net_pay
        }
        return ordered_result

    def process_dataframe(self, df):
        """Process entire dataframe of employees."""
        results = []
        for _, row in df.iterrows():
            results.append(self.process_employee(row.to_dict()))
        return pd.DataFrame(results)
    

File: js/salary-calculator.js
/**
 * Salary Calculator class for Nigerian payroll processing
 */
class SalaryCalculator {
    constructor(components) {
        this.components = components;
    }

    /**
     * Calculate monthly gross from annual gross
     * @param {number} annualGross - Annual gross salary
     * @returns {number} Monthly gross salary
     */
    calculateMonthlyGross(annualGross) {
        return Number((annualGross / 12).toFixed(2));
    }

    /**
     * Calculate individual salary components
     * @param {number} monthlyGross - Monthly gross salary
     * @param {number} workingDaysRatio - Ratio of days worked
     * @returns {Object} Salary components
     */
    calculateComponents(monthlyGross, workingDaysRatio = 1) {
        const components = {};
        for (const [component, percentage] of Object.entries(this.components)) {
            components[component] = Number(
                (monthlyGross * (percentage / 100) * workingDaysRatio).toFixed(2)
            );
        }
        return components;
    }

    /**
     * Calculate pension contributions
     * @param {Object} params - Pension calculation parameters
     * @returns {Object} Pension details
     */
    calculatePension({
        basic,
        transport,
        housing,
        contractType,
        monthlyGross,
        voluntaryPension = 0
    }) {
        if (contractType.trim().toUpperCase() === 'CONTRACT' || monthlyGross < 30000) {
            return {
                employeePension: 0,
                employerPension: 0,
                voluntaryPension: 0,
                totalPension: 0
            };
        }

        const pensionableBase = basic + transport + housing;
        const employeePension = Number((0.08 * pensionableBase).toFixed(2));
        const employerPension = Number((0.10 * pensionableBase).toFixed(2));
        const voluntary = Number(voluntaryPension.toFixed(2));
        const totalPension = employeePension + employerPension + voluntary;

        return {
            employeePension,
            employerPension,
            voluntaryPension: voluntary,
            totalPension
        };
    }

    /**
     * Calculate Consolidated Relief Allowance (CRA)
     * @param {number} grossPay - Gross pay
     * @param {number} pension - Pension deduction
     * @returns {number} CRA amount
     */
    calculateCRA(grossPay, pension) {
        const grossAfterPension = grossPay - pension;
        const craPercentage = Number((0.2 * grossAfterPension).toFixed(2));
        const minimumRelief = Number(
            Math.max(0.01 * grossAfterPension, 200000 / 12).toFixed(2)
        );
        return Number((craPercentage + minimumRelief).toFixed(2));
    }

    /**
     * Calculate PAYE tax
     * @param {number} taxablePay - Taxable pay amount
     * @returns {number} PAYE tax amount
     */
    calculatePAYE(taxablePay) {
        const taxBands = [
            [300000, 0.07],
            [300000, 0.11],
            [500000, 0.15],
            [500000, 0.19],
            [1600000, 0.21],
            [Infinity, 0.24]
        ];

        const annualTaxable = taxablePay * 12;
        let totalTax = 0;
        let remainingIncome = annualTaxable;

        for (const [band, rate] of taxBands) {
            if (remainingIncome <= 0) break;
            const taxableInBand = Math.min(band, remainingIncome);
            totalTax += taxableInBand * rate;
            remainingIncome -= band;
        }

        return Number((totalTax / 12).toFixed(2));
    }

    /**
     * Process salary calculations for a single employee
     * @param {Object} employee - Employee data
     * @returns {Object} Calculated salary details
     */
    processEmployee(employee) {
        // Calculate working days ratio
        const workingRatio = utils.calculateWorkingDaysRatio(
            employee['START DATE'],
            employee['END DATE']
        );

        // Calculate monthly gross
        const monthlyGross = this.calculateMonthlyGross(employee['ANNUAL GROSS PAY']);
        const proratedMonthlyGross = Number((monthlyGross * workingRatio).toFixed(2));

        // Get reimbursements and other deductions
        const reimbursements = Number(employee['Reimbursements'] || 0);
        const otherDeductions = Number(employee['Other Deductions'] || 0);

        // Calculate prorated components
        const components = this.calculateComponents(monthlyGross, workingRatio);

        // Calculate pension contributions
        const pensionDetails = this.calculatePension({
            basic: components.BASIC,
            transport: components.TRANSPORT,
            housing: components.HOUSING,
            contractType: employee['Contract Type'],
            monthlyGross: proratedMonthlyGross,
            voluntaryPension: Number(employee['VOLUNTARY_PENSION'] || 0)
        });

        // Calculate adjusted gross income for CRA
        const statutoryDeductions = pensionDetails.employeePension + pensionDetails.voluntaryPension;
        const adjustedGross = proratedMonthlyGross - statutoryDeductions;

        // Calculate CRA
        const cra = this.calculateCRA(adjustedGross, 0);

        // Calculate taxable pay
        const taxablePay = Number((adjustedGross - cra).toFixed(2));

        // Calculate PAYE tax
        const payeTax = this.calculatePAYE(taxablePay);

        // Calculate total deductions and net pay
        const totalDeductions = Number((
            payeTax +
            pensionDetails.employeePension +
            pensionDetails.voluntaryPension +
            otherDeductions
        ).toFixed(2));

        const netPay = Number((
            proratedMonthlyGross - totalDeductions + reimbursements
        ).toFixed(2));

        // Calculate total tax relief
        const totalTaxRelief = Number((
            cra +
            pensionDetails.employeePension +
            pensionDetails.voluntaryPension
        ).toFixed(2));

        return {
            'Account Number': employee['Account Number'],
            'STAFF ID': employee['STAFF ID'],
            'Email': employee['Email'],
            'NAME': employee['NAME'],
            'DEPARTMENT': employee['DEPARTMENT'],
            'JOB TITLE': employee['JOB TITLE'],
            'Contract Type': employee['Contract Type'],
            'ANNUAL GROSS PAY': employee['ANNUAL GROSS PAY'],
            'MONTHLY_GROSS': monthlyGross,
            'START DATE': employee['START DATE'],
            'END DATE': employee['END DATE'],
            'WORKING_DAYS_RATIO': workingRatio,
            'PRORATED_MONTHLY_GROSS': proratedMonthlyGross,
            'COMP_BASIC': components.BASIC,
            'COMP_TRANSPORT': components.TRANSPORT,
            'COMP_HOUSING': components.HOUSING,
            'COMP_UTILITY': components.UTILITY,
            'CRA': cra,
            'MANDATORY_PENSION': pensionDetails.employeePension,
            'VOLUNTARY_PENSION': pensionDetails.voluntaryPension,
            'EMPLOYER_PENSION': pensionDetails.employerPension,
            'TAX_RELIEF': totalTaxRelief,
            'TAXABLE_PAY': taxablePay,
            'PAYE_TAX': payeTax,
            'OTHER_DEDUCTIONS': otherDeductions,
            'REIMBURSEMENTS': reimbursements,
            'TOTAL_DEDUCTIONS': totalDeductions,
            'NET_PAY': netPay
        };
    }

    /**
     * Process salary calculations for multiple employees
     * @param {Object[]} employees - Array of employee data
     * @returns {Object[]} Array of calculated salary details
     */
    processDataframe(employees) {
        return employees.map(employee => this.processEmployee(employee));
    }
}

// Make SalaryCalculator available globally
window.SalaryCalculator = SalaryCalculator;
